File: map/lab2/src/repository/StudentValidator.java
package repository;

import domain.Student;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class StudentValidator implements Validator<Student> {
    /**
     * Validates if a string is email type.
     *
     * @param email
     *   Param to validate.
     *
     * @return
     *   Returns true if variable email is of email format.
     */
    private boolean validateEmail(String email) {
        String ePattern = "^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@((\\[[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}])|(([a-zA-Z\\-0-9]+\\.)+[a-zA-Z]{2,}))$";
        Pattern p = Pattern.compile(ePattern);
        Matcher m = p.matcher(email);

        return m.matches();
    }

    /**
     * Validates if a string is a name.
     *
     * @param name
     *   The name to check
     *
     * @return
     *   Returns true if variable name is a name.
     */
    private boolean validateName(String name) {
        String ePattern = "^[a-zA-Z]+ [a-zA-Z]+[- ]?[a-zA-Z]+$";
        Pattern p = Pattern.compile(ePattern);
        Matcher m = p.matcher(name);

        return m.matches();
    }

    /**
     * Validates variable st.
     * @param st
     *   st - student to validate.
     *
     * @throws ValidationException
     *   Throws ValidationException if the validation fails.
     */
    @Override
    public void validate(Student st) throws ValidationException {
        if (!validateEmail(st.getEmail()))
            throw new ValidationException("The email is invalid.");

        if (!validateName(st.getName()))
            throw new ValidationException("The name is invalid.");

        if (!validateName(st.getGuide()))
            throw new ValidationException("The guide's name is invalid.");

        if (st.getId() < 0)
            throw new ValidationException("The id is invalid.");
    }
}

File: map/lab7/lab7/Domain/StudentValidator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace lab7.Domain
{
    public class StudentValidator : IValidator<Student>
    {
        private bool ValidateEmail(string email) {
            Regex p = new Regex("^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@((\\[[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}])|(([a-zA-Z\\-0-9]+\\.)+[a-zA-Z]{2,}))$");
            Match m = p.Match(email);

            return m.Success;
        }
        
        private bool ValidateName(string name)
        {
            var p = new Regex("^[a-zA-Z]+ [a-zA-Z]+[- ]?[a-zA-Z]+$");
            Match m = p.Match(name);

            return m.Success;
        }
        
        public void Validate(Student st)
        {
            var errorMsg = new List<string>();
            
            if (st.Id <= 0)
                errorMsg.Add("The id is invalid.");

            if (!ValidateName(st.Name))
                errorMsg.Add("The name is invalid.");

            if (st.Group.Length != 3)
                errorMsg.Add("The group is invalid.");

            if (!ValidateEmail(st.Email))
                errorMsg.Add("The email is invalid.");

            if (!ValidateName(st.Guide))
                errorMsg.Add("The guide's name is invalid.");

            if (errorMsg.Any())
                throw new ValidationException(String.Join("\n", errorMsg.ToArray()));
        }
    }
}